### PR TITLE
build: integrate Python into build system

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Let's move the `/pybridge` scenario over to building the Python bridge as part of the normal RPM build process.

 - [x] cockpit-project/bots#4707
 - [x] consider waiting until after the 291 release

### CODE REVIEW

**Comment by martinpitt** [comment]
@allisonkarlitskaya Do you still plan to actually exercise this in CI somewhere? Right now that is totally dead code. It would be nice to replace the `--python` hack in test/image-prepare with something like defining an RPM macro. Want some help with that?

**Comment by martinpitt** [comment]
I checked
```
./autogen.sh --prefix=/usr --with-cockpit-user=cockpit-ws --with-cockpit-ws-instance-user=cockpit-wsinstance --sysconfdir=/etc --localstatedir=/var --enable-strict --enable-python
make install DESTDIR=/tmp/i
```
.. and it looks mostly good. I found two remaining issues:

 - `cockpit-askpass` ends up in /usr/bin instead of /usr/libexec; needs to be moved there (in the Makefile rule after pip install, as AFAIK setuptools doesn't know about libexecdir?), and the packages.py patch hack reverted
 - Still installs /usr/libexec/cockpit-ssh -- this isn't necessary any more with ferny, and causes a libssh dependency. This might be as simple as `--enable-python` implying `--disable-ssh`?

Let's coordinate the next steps in the chat today. Thanks!

**Comment by allisonkarlitskaya** [comment]
The `/pybridge` fails were expected, but the dpkg failures are most unwelcome...

**Comment by martinpitt** [comment]
The bots PR landed. Wrt. [fedora-37](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18723-20230429-014548-13131878-fedora-37-pybridge-bots-4707/log.html), the stratis failure got naughtied, test-loopback needs looking at. That failed in c8s as well. Could be a flake, but maybe also some subtle change.

The other [c8s failures](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18723-20230429-014542-13131878-centos-8-stream-pybridge-bots-4707/log.html#67) are all the same unexpected message -- @allisonkarlitskaya perhaps you can have a quick look if it's an obvious one, otherwise we can just add it to the pile of expected messages for now (not important enough to block this PR).

File: configure.ac
Comment (martinpitt) [inline configure.ac:79]: I find this rather confusing -- this doesn't disable the bridge at all, there will still be a /usr/bin/cockpit-bridge . What happened to `--enable-pybridge` or `-python`?

File: pyproject.toml
Comment (martinpitt) [inline pyproject.toml]: This looks like a rebase error -- that looks like it should be in the "add tox support" commit, not in the "tweak version number scheme" one?

File: setup.cfg
Comment (martinpitt) [inline setup.cfg:8]: That looks really strange -- I understand that there can be a newline and indentation after `=` to have a multi-line value, but the repeated `=` looks like an error?

File: pkg/Makefile.am
Comment (martinpitt) [inline pkg/Makefile.am:46]: Can you please add a `FIXME:` comment here? The qunit tests are very much supposed to work against the python bridge, and we have relied on them heavily to ensure feature parity and for fixing bugs. I'm ok with temporarily disabling them to unblock this, but for me it would then be a #1 priority to get them back.

File: src/bridge/Makefile.am
Comment (martinpitt) [inline src/bridge/Makefile.am:119]: Note that this still builds libcockpit-bridge.a (and all of the C files it contains). Not a biggie, but should also be rather easy to fix?

File: tools/debian/rules
Comment (martinpitt) [inline tools/debian/rules:13]: Instead of the clean hack below, try to add `--buildsystem=autoconf` here. That should skip the auto-detection and preferring setup.py.

File: tools/debian/rules
Comment (martinpitt) [inline tools/debian/rules:17]: If the `--buildsystem` above does not work, please do this as
```
execute_before_dh_clean:
        rm -f setup.py
```
but this would still be rather nasty, as it will at least cause a warning during source package build that an original file got removed.

File: tools/cockpit.spec
Comment (martinpitt) [inline tools/cockpit.spec:228]: Both of this needs to happen in Makefile.am, this knowledge does not belong into the spec. It would have to be repeated in the Debian, arch, gentoo, etc. packaging, and breaks `make install` -- that currently doesn't install /usr/bin/cockpit-bridge or c-askpass at all.

I know that this will install the py bridge for !build_basic as well, but we already have to filter out files for the C bridge anyway.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Whee, I'm glad this is coming together! Thanks for doing the heavy lifting with image builds etc. already.

File: configure.ac
Comment (allisonkarlitskaya) [reply configure.ac:79]: From our discussion: I'll switch this back to `--enable-python`, move the python installation (and moving askpass to libexecdir) back into `make install`, switch to a PEP517 install, and remove `setup.py`.

File: pyproject.toml
Comment (allisonkarlitskaya) [reply pyproject.toml]: Thanks for the catch.  `xdist` also doesn't really belong in this list — we don't actually use it.

File: setup.cfg
Comment (allisonkarlitskaya) [reply setup.cfg:8]: This is copied verbatim from the setuptools documentation for src-layout packages: https://setuptools.pypa.io/en/latest/userguide/package_discovery.html

From what I understand, the `package_dir=` attribute should be read as a multi-line mapping.  In this case our mapping contains a single item which maps the empty string to `src`.

The syntax is absolutely bizarre, but this is stock usage, so I don't think we need to document it.

File: pkg/Makefile.am
Comment (allisonkarlitskaya) [reply pkg/Makefile.am:46]: I did this for two reasons:
 - we run the tests as part of pytest anyway
 - the tests don't work anyway since there's no `./cockpit-bridge` in the build directory for us to run.  Even when I tried to create it, something wasn't working right.  I didn't spend much more time trying to sort it out because, as written above, we run them as part of pytest anyway.

File: src/bridge/Makefile.am
Comment (allisonkarlitskaya) [reply src/bridge/Makefile.am:119]: I looked into this.  It's not easy.  In order to get this treated as a library we need to introduce it somehow, which we do via a `noinst_` target.  Those get built as part of `make all` even if they're not depended on by anything.

A quick look through the docs didn't turn up a way to get a `nobuild_noinst_` target...

We could do some more conditionals here to avoid building these, but it would be a little bit complicated: we want to build `libcockpit-bridge.a` even in the `--disable-bridge` case for `cockpit-pcp`.  If we had `--disable-bridge` AND `--disable-pcp` then we could probably exclude this, and also the related metrics library, but, honestly, I didn't think it was worth it for the added complexity.

File: tools/debian/rules
Comment (allisonkarlitskaya) [reply tools/debian/rules:13]: YES!  This is what I was looking for, thanks.

File: tools/debian/rules
Comment (allisonkarlitskaya) [reply tools/debian/rules:17]: I also agree that this would be better than what I did.

File: tools/cockpit.spec
Comment (allisonkarlitskaya) [reply tools/cockpit.spec:228]: See above comment about this.

File: src/bridge/Makefile.am
Comment (martinpitt) [reply src/bridge/Makefile.am:119]: Agreed, this isn't worth wasting time on. I just wanted to make sure it's not a simple oversight.

File: setup.cfg
Comment (martinpitt) [reply setup.cfg:8]: Thanks for the heads-up!

File: tools/debian/rules
Comment (allisonkarlitskaya) [reply tools/debian/rules:13]: It worked.

File: tools/debian/rules
Comment (allisonkarlitskaya) [reply tools/debian/rules:17]: Not doing this one, since the above idea worked.

**Comment by martinpitt** [comment]
I can reproduce the [loopback test failure](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18723-20230429-014548-13131878-fedora-37-pybridge-bots-4707/log.html#348-1). Just as the CI run did ("3x affected"), running the test twice in the same VM fails. It seems to fail in about half of the runs here, but there's no clear pattern. I'll keep this comment updated after more debugging (and lunch/meeting first).

One thing I noticed: This still installs /usr/libexec/cockpit-ssh, and that is actually running during this test. I think direct remote logins from cockpit-ws weren't ported to ferny yet.

File: src/bridge/Makefile.am
Comment (allisonkarlitskaya) [reply src/bridge/Makefile.am:119]: I actually took another look at this: it's `libcockpit-metrics.a` which is used by both `cockpit-pcp` and `cockpit-bridge` and would be difficult to conditionalize.  `libcockpit-bridge.a` is only used by the bridge and its tests and can very easily be tucked into the conditional.  Fixed that.

File: tools/cockpit.spec
Comment (martinpitt) [inline tools/cockpit.spec:295]: Needs `-r`, see [failed log](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18723-20230502-142448-c9871460-centos-8-stream/log.html)

Comment (allisonkarlitskaya) [reply tools/cockpit.spec:295]: arghgh

**Comment by martinpitt** [comment]
[This testCockpitDesktop](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18723-20230502-152744-0811c950-centos-8-stream-pybridge/log.html#16-2) failure is because the process is now

    /usr/libexec/platform-python /usr/bin/cockpit-bridge --privileged

which the RE doesn't match. Tested fix:

```diff
--- test/verify/check-connection
+++ test/verify/check-connection
@@ -889,7 +889,8 @@ ResultInactive=yes
 ResultActive=yes""")
 
         if os.environ.get("TEST_SCENARIO") == "pybridge":
-            pgrep = "^/usr/bin/python.*cockpit-bridge.*--privileged"
+            # /usr/bin/python on Fedora, /usr/libexec/platform-python on RHEL
+            pgrep = "^/usr/[^[:space:]]*python.*cockpit-bridge.*--privileged"
         else:
             pgrep = "^/usr/bin/cockpit-bridge.*--privileged"
 
```

**Comment by allisonkarlitskaya** [comment]
I came up with my own fix for the RE:

```sh
until pgrep -f '^(/usr/[^ ]+/[^ /]*python[^ /]* )?/usr/bin/cockpit-bridge'; do sleep 1; done
```

It's one universal option which will match /usr/bin/cockpit-bridge as optionally prefixed by something in `/usr` which has `python` in its basename.  I'd like to remove as much `pybridge`-conditional-stuff from the integration tests as possible...

File: tools/cockpit.spec
Comment (martinpitt) [inline tools/cockpit.spec:56]: The commit message still says "We move cockpit-askpass from /usr/bin to /usr/libexec". That's no longer true for the spec, as Makefile already does it.

**Review by martinpitt — APPROVED** [review]
Thanks! This looks really good to me, and the commit message isn't enough reason to push this again. If this goes green, let's get it in!

**Review by martinpitt — CHANGES_REQUESTED** [review]
Ah, [meh](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18723-20230502-163851-2a54d501-centos-8-stream-pybridge/log.html#123) -- so these unexpected messages are still there, but are now *much* easier to handle, as they are a single line like

    cockpit.router-ERROR: trying to drop non-existent channel 1:3!37 from {}

**Update**: Done

**Review by martinpitt — DISMISSED** [review]

File: tools/cockpit.spec
Comment (martinpitt) [reply tools/cockpit.spec:56]: I dropped that paragraph.

**Comment by allisonkarlitskaya** [comment]
Blocking because maybe we don't want to toss this into the release at the last minute...

**Comment by allisonkarlitskaya** [comment]
Let's let the pilot decide.

@mvollmer ?

**Comment by mvollmer** [comment]
> Let's let the pilot decide.
> 
> @mvollmer ?

Let's get it in, and see if it breaks the release process. Better now than later.

**Review by martinpitt — APPROVED** [review]

**Pull request merged**
